From d361e8ce8d81607e8b14633004e94e33f1742fc0 Mon Sep 17 00:00:00 2001
From: Volodymyr Pavlenko <pavlenko.vv@pm.me>
Date: Sun, 13 Apr 2025 20:12:53 +0300
Subject: [PATCH 2/7] Controller: add service_prioritization_task

- handling of QOS_MANAGEMENT_NOTIFICATION_MESSAGE
- workflow for QoS management descriptors

Signed-off-by: Volodymyr Pavlenko <pavlenko.vv@pm.me>
---
 controller/src/beerocks/master/CMakeLists.txt |   1 +
 controller/src/beerocks/master/controller.cpp |  73 +---------
 controller/src/beerocks/master/controller.h   |   2 -
 .../tasks/service_prioritization_task.cpp     | 136 ++++++++++++++++++
 .../tasks/service_prioritization_task.h       |  55 +++++++
 5 files changed, 197 insertions(+), 70 deletions(-)
 create mode 100644 controller/src/beerocks/master/tasks/service_prioritization_task.cpp
 create mode 100644 controller/src/beerocks/master/tasks/service_prioritization_task.h

diff --git a/controller/src/beerocks/master/CMakeLists.txt b/controller/src/beerocks/master/CMakeLists.txt
index 3de0c34a7..0e2122af8 100644
--- a/controller/src/beerocks/master/CMakeLists.txt
+++ b/controller/src/beerocks/master/CMakeLists.txt
@@ -71,6 +71,7 @@ endif()
 target_link_libraries(${PROJECT_NAME} ${LINKED_LIBS})
 
 target_include_directories(${PROJECT_NAME} PRIVATE
+    ${CMAKE_CURRENT_SOURCE_DIR}
     ${MODULE_PATH}/../bml
     if (ENABLE_NBAPI)
         ${ACTION_PATH}
diff --git a/controller/src/beerocks/master/controller.cpp b/controller/src/beerocks/master/controller.cpp
index 2eb379940..1afc2b897 100644
--- a/controller/src/beerocks/master/controller.cpp
+++ b/controller/src/beerocks/master/controller.cpp
@@ -19,6 +19,7 @@
 #include "tasks/dhcp_task.h"
 #include "tasks/load_balancer_task.h"
 #include "tasks/optimal_path_task.h"
+#include "tasks/service_prioritization_task.h"
 #include "tasks/topology_task.h"
 #ifdef FEATURE_PRE_ASSOCIATION_STEERING
 #include "tasks/pre_association_steering/pre_association_steering_task.h"
@@ -386,6 +387,10 @@ void Controller::start_mandatory_tasks()
 
     LOG_IF(!m_task_pool.add_task(std::make_shared<DhcpTask>(database, m_timer_manager)), FATAL)
         << "Failed adding dhcp task!";
+
+    LOG_IF(!m_task_pool.add_task(std::make_shared<service_prioritization_task>(database, cmdu_tx)),
+           FATAL)
+        << "Failed adding service_prioritization_task!";
 }
 
 void Controller::start_optional_tasks()
@@ -607,8 +612,6 @@ bool Controller::handle_cmdu_1905_1_message(const sMacAddr &src_mac,
         return handle_cmdu_1905_failed_connection_message(src_mac, cmdu_rx);
     case ieee1905_1::eMessageType::ASSOCIATED_STA_LINK_METRICS_RESPONSE_MESSAGE:
         return handle_cmdu_1905_associated_sta_link_metrics_response_message(src_mac, cmdu_rx);
-    case ieee1905_1::eMessageType::QOS_MANAGEMENT_NOTIFICATION_MESSAGE:
-        return handle_cmdu_1905_qos_management_notification_message(src_mac, cmdu_rx);
     case ieee1905_1::eMessageType::EARLY_AP_CAPABILITY_REPORT_MESSAGE:
         return handle_cmdu_1905_early_ap_capability_report_message(src_mac, cmdu_rx);
 
@@ -4668,72 +4671,6 @@ bool Controller::handle_tlv_profile3_device_inventory(Agent &agent,
     return true;
 }
 
-bool Controller::handle_cmdu_1905_qos_management_notification_message(
-    const sMacAddr &src_mac, ieee1905_1::CmduMessageRx &cmdu_rx)
-{
-    auto mid = cmdu_rx.getMessageId();
-    LOG(DEBUG) << "Received QOS_MANAGEMENT_NOTIFICATION_MESSAGE, mid=" << std::hex << mid;
-
-    for (auto const &qos_management_descriptor_tlv :
-         cmdu_rx.getClassList<wfa_map::tlvQoSManagementDescriptor>()) {
-        if (!qos_management_descriptor_tlv) {
-            LOG(DEBUG) << "getClass wfa_map::tlvQoSManagementDescriptor has failed";
-            return false;
-        }
-
-        LOG(DEBUG) << "QoS Management Descriptor TLV is received:" << std::endl
-                   << "Rule ID: " << qos_management_descriptor_tlv->qmid() << std::endl
-                   << "BSSID: " << qos_management_descriptor_tlv->bssid() << std::endl
-                   << "Client MAC: " << qos_management_descriptor_tlv->client_mac() << std::endl;
-
-        // TODO: Implement handling of QoS Management Descriptor TLV (PPM-2364)
-    }
-
-    auto agent = database.m_agents.get(src_mac);
-    if (!agent) {
-        LOG(ERROR) << "Agent with mac is not found in database mac=" << src_mac;
-        return false;
-    }
-
-    if (!cmdu_tx.create(0, ieee1905_1::eMessageType::SERVICE_PRIORITIZATION_REQUEST_MESSAGE)) {
-        LOG(ERROR) << "cmdu creation of type SERVICE_PRIORITIZATION_REQUEST_MESSAGE has failed";
-        return false;
-    }
-
-    for (const auto &rule : agent->service_prioritization.rules) {
-
-        auto service_prioritization_rule_tlv =
-            cmdu_tx.addClass<wfa_map::tlvServicePrioritizationRule>();
-        if (!service_prioritization_rule_tlv) {
-            LOG(ERROR) << "addClass wfa_map::tlvServicePrioritizationRule has failed";
-            return false;
-        }
-
-        service_prioritization_rule_tlv->rule_params() = rule.second;
-    }
-
-    auto dscp_mapping_table_tlv = cmdu_tx.addClass<wfa_map::tlvDscpMappingTable>();
-    if (!dscp_mapping_table_tlv) {
-        LOG(ERROR) << "addClass wfa_map::tlvDscpMappingTable has failed";
-        return false;
-    }
-
-    if (!dscp_mapping_table_tlv->set_dscp_pcp_mapping(
-            agent->service_prioritization.dscp_mapping_table.data(),
-            agent->service_prioritization.dscp_mapping_table.size())) {
-        LOG(ERROR) << "Failed to set DSCP mapping list in tlvDscpMappingTable!";
-        return false;
-    }
-
-    if (!database.dm_set_service_prioritization_rules(*agent)) {
-        LOG(ERROR) << "Failed to set service prioritization rules in DM for Agent" << agent->al_mac;
-        return false;
-    }
-
-    // TODO: Implement sending of remaining TLVs of Service Prioritization Request message (PPM-2366)
-    return son_actions::send_cmdu_to_agent(src_mac, cmdu_tx, database);
-}
-
 bool Controller::handle_ap_capability_report(const sMacAddr &src_mac,
                                              ieee1905_1::CmduMessageRx &cmdu_rx, const bool early)
 {
diff --git a/controller/src/beerocks/master/controller.h b/controller/src/beerocks/master/controller.h
index 73a44199a..91763fd03 100644
--- a/controller/src/beerocks/master/controller.h
+++ b/controller/src/beerocks/master/controller.h
@@ -373,8 +373,6 @@ private:
         const sMacAddr &src_mac, ieee1905_1::CmduMessageRx &cmdu_rx);
     bool handle_cmdu_1905_bss_configuration_request_message(const sMacAddr &src_mac,
                                                             ieee1905_1::CmduMessageRx &cmdu_rx);
-    bool handle_cmdu_1905_qos_management_notification_message(const sMacAddr &src_mac,
-                                                              ieee1905_1::CmduMessageRx &cmdu_rx);
 
     bool handle_ap_capability_report(const sMacAddr &src_mac, ieee1905_1::CmduMessageRx &cmdu_rx,
                                      const bool early);
diff --git a/controller/src/beerocks/master/tasks/service_prioritization_task.cpp b/controller/src/beerocks/master/tasks/service_prioritization_task.cpp
new file mode 100644
index 000000000..28b9dfd2c
--- /dev/null
+++ b/controller/src/beerocks/master/tasks/service_prioritization_task.cpp
@@ -0,0 +1,136 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include "service_prioritization_task.h"
+
+#include "son_actions.h"
+
+#include <tlvf/wfa_map/tlvDscpMappingTable.h>
+#include <tlvf/wfa_map/tlvQoSManagementDescriptor.h>
+#include <tlvf/wfa_map/tlvQoSManagementPolicy.h>
+#include <tlvf/wfa_map/tlvServicePrioritizationRule.h>
+
+namespace son {
+
+service_prioritization_task::service_prioritization_task(db &database_,
+                                                         ieee1905_1::CmduMessageTx &cmdu_tx_,
+                                                         const std::string &task_name_)
+    : task(task_name_), m_db(database_), m_cmdu_tx(cmdu_tx_)
+{
+}
+
+bool service_prioritization_task::handle_ieee1905_1_msg(const sMacAddr &src_mac,
+                                                        ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    switch (cmdu_rx.getMessageType()) {
+    case ieee1905_1::eMessageType::QOS_MANAGEMENT_NOTIFICATION_MESSAGE:
+        return handle_cmdu_1905_qos_management_notification_message(src_mac, cmdu_rx);
+    default: {
+        return false;
+    }
+    }
+}
+
+bool service_prioritization_task::handle_cmdu_1905_qos_management_notification_message(
+    const sMacAddr &src_mac, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received QOS_MANAGEMENT_NOTIFICATION_MESSAGE, mid=" << std::hex << mid;
+
+    // --- Prepare SERVICE_PRIORITIZATION_REQUEST_MESSAGE to be sent to all agents ---
+    if (!m_cmdu_tx.create(0, ieee1905_1::eMessageType::SERVICE_PRIORITIZATION_REQUEST_MESSAGE)) {
+        LOG(ERROR) << "cmdu creation of type SERVICE_PRIORITIZATION_REQUEST_MESSAGE has failed";
+        return false;
+    }
+
+    // --- Handle Zero or more Service Prioritization Rule TLVs ---
+    auto agent = m_db.m_agents.get(src_mac);
+    if (!agent) {
+        LOG(ERROR) << "Agent with mac is not found in database mac=" << src_mac;
+        return false;
+    }
+
+    for (const auto &rule : agent->service_prioritization.rules) {
+
+        auto service_prioritization_rule_tlv =
+            m_cmdu_tx.addClass<wfa_map::tlvServicePrioritizationRule>();
+        if (!service_prioritization_rule_tlv) {
+            LOG(ERROR) << "addClass wfa_map::tlvServicePrioritizationRule has failed";
+            return false;
+        }
+
+        service_prioritization_rule_tlv->rule_params() = rule.second;
+    }
+
+    // --- Handle Zero or one DSCP Mapping Table TLV  ---
+    auto dscp_mapping_table_tlv = m_cmdu_tx.addClass<wfa_map::tlvDscpMappingTable>();
+    if (!dscp_mapping_table_tlv) {
+        LOG(ERROR) << "addClass wfa_map::tlvDscpMappingTable has failed";
+        return false;
+    }
+
+    if (!dscp_mapping_table_tlv->set_dscp_pcp_mapping(
+            agent->service_prioritization.dscp_mapping_table.data(),
+            agent->service_prioritization.dscp_mapping_table.size())) {
+        LOG(ERROR) << "Failed to set DSCP mapping list in tlvDscpMappingTable!";
+        return false;
+    }
+
+    if (!m_db.dm_set_service_prioritization_rules(*agent)) {
+        LOG(ERROR) << "Failed to set service prioritization rules in DM for Agent" << agent->al_mac;
+        return false;
+    }
+
+    // --- Handle Zero or more QoS Management Descriptor TLVs ---
+    for (auto const &qos_management_descriptor_tlv_in :
+         cmdu_rx.getClassList<wfa_map::tlvQoSManagementDescriptor>()) {
+        if (!qos_management_descriptor_tlv_in) {
+            LOG(DEBUG) << "getClass wfa_map::tlvQoSManagementDescriptor has failed";
+            return false;
+        }
+
+        LOG(DEBUG) << "QoS Management Descriptor TLV is received:" << std::endl
+                   << "Rule ID: " << qos_management_descriptor_tlv_in->qmid() << std::endl
+                   << "BSSID: " << qos_management_descriptor_tlv_in->bssid() << std::endl
+                   << "Client MAC: " << qos_management_descriptor_tlv_in->client_mac() << std::endl;
+
+        m_descriptors[m_qmid_next] = qos_management_descriptor_tlv_in;
+
+        // forwarding the QoS Management Descriptor TLVs to the agent with specified QMIDs
+        auto qos_management_descriptor_tlv_out =
+            m_cmdu_tx.addClass<wfa_map::tlvQoSManagementDescriptor>();
+        if (!qos_management_descriptor_tlv_out) {
+            LOG(ERROR) << "addClass wfa_map::tlvQoSManagementDescriptor for output has failed";
+            return false;
+        }
+        qos_management_descriptor_tlv_out->bssid() = qos_management_descriptor_tlv_in->bssid();
+        qos_management_descriptor_tlv_out->client_mac() =
+            qos_management_descriptor_tlv_in->client_mac();
+        qos_management_descriptor_tlv_out->qmid() = m_qmid_next;
+        qos_management_descriptor_tlv_out->set_descriptor_element(
+            qos_management_descriptor_tlv_in->descriptor_element(),
+            qos_management_descriptor_tlv_in->descriptor_element_length());
+
+        m_qmid_next++;
+    }
+
+    // TODO: PPM-3268: Check for agent SCS/MSCS capabilities before sending
+    bool result = true;
+    for (const auto &agent_entry : m_db.m_agents) {
+        result = son_actions::send_cmdu_to_agent(agent_entry.first, m_cmdu_tx, m_db);
+        if (!result) {
+            LOG(ERROR) << "Failed to send SERVICE_PRIORITIZATION_REQUEST_MESSAGE to agent "
+                       << agent_entry.first;
+            result = false;
+        }
+    }
+
+    return result;
+}
+
+} // namespace son
diff --git a/controller/src/beerocks/master/tasks/service_prioritization_task.h b/controller/src/beerocks/master/tasks/service_prioritization_task.h
new file mode 100644
index 000000000..1e2c629d8
--- /dev/null
+++ b/controller/src/beerocks/master/tasks/service_prioritization_task.h
@@ -0,0 +1,55 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2025 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef _SERVICE_PRIORITIZATION_TASK_H_
+#define _SERVICE_PRIORITIZATION_TASK_H_
+
+#include "task.h"
+
+#include <db/db.h>
+
+#include <memory>
+
+namespace wfa_map {
+class tlvQoSManagementDescriptor;
+};
+
+namespace son {
+
+class service_prioritization_task : public task {
+
+public:
+    service_prioritization_task(db &database_, ieee1905_1::CmduMessageTx &cmdu_tx_,
+                                const std::string &task_name_ = std::string());
+    virtual ~service_prioritization_task() {}
+
+    bool handle_ieee1905_1_msg(const sMacAddr &src_mac,
+                               ieee1905_1::CmduMessageRx &cmdu_rx) override;
+
+protected:
+    void work() override{};
+
+private:
+    bool handle_cmdu_1905_qos_management_notification_message(const sMacAddr &src_mac,
+                                                              ieee1905_1::CmduMessageRx &cmdu_rx);
+
+private:
+    db &m_db;
+    ieee1905_1::CmduMessageTx &m_cmdu_tx;
+    std::unordered_map<uint16_t, std::shared_ptr<wfa_map::tlvQoSManagementDescriptor>>
+        m_descriptors;
+
+    /**
+     * @brief next number of QMID to be used, incrementing after usage is required
+     **/
+    uint64_t m_qmid_next{};
+};
+
+} // namespace son
+
+#endif // _SERVICE_PRIORITIZATION_TASK_H_
-- 
2.34.1

